src/watch_state.rs

```diff
@@ -0,0 +1,231 @@
+use std::sync::{Arc, Mutex};
+
+use watchexec::action::Outcome;
+
+/// Tracks the current state of spin watch. Handles transitioning states and converting effects
+/// into outcomes based on the current state.
+#[derive(Debug, Clone)]
+pub struct StateMachine {
+    state: Arc<Mutex<State>>,
+    clear: bool,
+    skip_build: bool,
+}
+
+impl StateMachine {
+    pub fn new(skip_build: bool, clear: bool) -> Self {
+        Self {
+            state: Arc::new(Mutex::new(match skip_build {
+                false => State::Building,
+                true => State::Running,
+            })),
+            clear,
+            skip_build,
+        }
+    }
+
+    /// Get the current state of the state machine.
+    pub fn get_state(&self) -> State {
+        *self.state.lock().unwrap()
+    }
+
+    /// Based on the given effect return the correct outcome. Transition the internal state of the
+    /// machine if necessary.
+    pub fn handle(&self, effect: Effect) -> Outcome {
+        let mut state = self.state.lock().unwrap();
+        tracing::debug!("handling effect {:?} in current state {:?}", effect, *state);
+        let outcome = match (effect, *state) {
+            (Effect::Exit, _) => {
+                Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
+            }
+            (Effect::ProcessFailed, _) => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
```

Comment:
I am struggling here a bit with `Outcome::if_running` and `State::Running`.  Does `if_running` refer to the `State`, or to something that is happening within the state machine itself?

Reply:
`Outcome::if_running` refers to [this](https://docs.rs/watchexec/latest/watchexec/action/enum.Outcome.html). Notably it says:

> Logic against the state of the command should be expressed using these variants, rather than inside the action handler, as it ensures the state of the command is always the latest available when the outcome is executed.

`State::Running` means we believe that `spin up` is currently running or at least we want it to be.

Basically wrapping all the outcomes in `Outcome::if_running` protects us from a scenario in which somehow our internal state gets out of sync with the actual state of watchexec. Likely not to matter but good protection regardless.

I added a comment to try and add some clarity.